cpp/src/io/parquet/page_enc.cu

```diff
@@ -0,0 +1,1550 @@
+    cur_fld = f;                                    \
+
+#define CPW_FLD_INT32(f, v)                         \
+    p = cpw_put_fldh(p, f, cur_fld, ST_FLD_I32);    \
+    p = cpw_put_int32(p, v);                        \
+    cur_fld = f;                                    \
+
+#define CPW_FLD_INT64(f, v)                         \
+    p = cpw_put_fldh(p, f, cur_fld, ST_FLD_I64);    \
+    p = cpw_put_int64(p, v);                        \
+    cur_fld = f;                                    \
+
+#define CPW_FLD_BINARY(f, v, l)                     \
+    p = cpw_put_fldh(p, f, cur_fld, ST_FLD_BINARY); \
+    p = cpw_put_uint32(p, l);                       \
+    memcpy(p, v, l);                                \
+    p += l;                                         \
+    cur_fld = f;                                    \
+
+#define CPW_END_STRUCT(hdr_end)                     \
+    *p++ = 0;                                       \
+    hdr_end = p;                                    \
+    }                                               \
+
+#define CPW_END_STRUCT_NOTERMINATION(hdr_end)       \
+    hdr_end = p;                                    \
+    }                                               \
+
+
+__device__ uint8_t *EncodeStatistics(uint8_t *start, const statistics_chunk *s, const EncColumnDesc *col, float *fp_scratch)
+{
+    uint8_t *end, dtype, dtype_len;
+    dtype = col->stats_dtype;
+    switch (dtype) {
+    case dtype_bool8:
+    case dtype_int8:
+        dtype_len = 1;
+        break;
+    case dtype_int16:
+        dtype_len = 2;
+        break;
+    case dtype_int32:
+    case dtype_float32:
+        dtype_len = 4;
+        break;
+    case dtype_int64:
+    case dtype_timestamp64:
+    case dtype_float64:
+        dtype_len = 8;
+        break;
+    case dtype_int128:
+        dtype_len = 16;
+        break;
+    case dtype_string:
+    default:
+        dtype_len = 0;
+        break;
+    }
+    CPW_BEGIN_STRUCT(start)
+        CPW_FLD_INT64(3, s->null_count)
+        if (s->has_minmax) {
+            const void *vmin, *vmax;
+            uint32_t lmin, lmax;
+
+            if (dtype == dtype_string) {
+                lmin = s->min_value.str_val.length;
+                vmin = s->min_value.str_val.ptr;
+                lmax = s->max_value.str_val.length;
+                vmax = s->max_value.str_val.ptr;
+            }
+            else {
+                lmin = lmax = dtype_len;
+                if (dtype == dtype_float32) { // Convert from double to float32
+                    fp_scratch[0] = s->min_value.fp_val;
+                    fp_scratch[1] = s->max_value.fp_val;
+                    vmin = &fp_scratch[0];
+                    vmax = &fp_scratch[1];
+                }
+                else {
+                    vmin = &s->min_value;
+                    vmax = &s->max_value;
+                }
+            }
+            CPW_FLD_BINARY(5, vmin, lmin);
+            CPW_FLD_BINARY(6, vmax, lmax);
```

Comment:
The min and max values are backwards here.  5==max and 6==min according to [the official Parquet thrift specification](https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L223-L224).

Reply:
Yup, thanks for the quick investigation.